File: frontend/src/types/reasoner.ts
export interface ReasonerViolation {
  type: 'UnsatisfiableClass' | 'CardinalityViolation' | 'DisjointViolation' | 'DomainRangeViolation'
  subject_iri: string
  description: string
}

export interface InferredAxiom {
  subject: string
  predicate: string
  object: string
  inference_rule: string
}

export interface ReasonerResultData {
  consistent: boolean
  violations: ReasonerViolation[]
  inferred_axioms: InferredAxiom[]
  execution_ms: number
}

export interface ReasonerResult {
  job_id: string
  ontology_id: string
  status: 'pending' | 'running' | 'completed' | 'failed'
  created_at: string
  completed_at?: string
  result?: ReasonerResultData
  error?: string
}

export interface ReasonerJob {
  job_id: string
  status: 'pending' | 'running' | 'completed' | 'failed'
}

export interface ReasonerRunRequest {
  subgraph_entity_iris?: string[]
  include_inferences?: boolean
  check_consistency?: boolean
  reasoner_profile?: 'OWL_DL' | 'OWL_EL' | 'OWL_RL' | 'OWL_QL'
}

File: backend/models/reasoner.py
"""
Reasoner Pydantic models
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class ReasonerViolation(BaseModel):
    violation_type: str
    subject_iri: str
    predicate_iri: str | None = None
    object_iri: str | None = None
    message: str
    severity: Literal["error", "warning", "info"] = "error"


class InferredAxiom(BaseModel):
    subject_iri: str
    predicate_iri: str
    object_iri: str
    inference_rule: str
    confidence: float = 1.0


class ReasonerResult(BaseModel):
    ontology_id: str
    job_id: str
    status: Literal["pending", "running", "completed", "failed"]
    violations: list[ReasonerViolation] = []
    inferred_axioms: list[InferredAxiom] = []
    violation_count: int = 0
    inferred_count: int = 0
    elapsed_seconds: float | None = None
    error: str | None = None
    created_at: datetime
    completed_at: datetime | None = None


class ReasonerJob(BaseModel):
    job_id: str
    ontology_id: str
    status: Literal["pending", "running", "completed", "failed"]
    created_at: datetime
    updated_at: datetime


class ReasonerRunRequest(BaseModel):
    subgraph_iris: list[str] | None = None  # None이면 전체 온톨로지
    include_inferences: bool = True
    check_consistency: bool = True
    reasoner_profile: Literal["EL", "RL", "QL", "FULL"] = "EL"
